### PR TITLE
`Offline Entitlements`: created `LoadShedderIntegrationTests`

### PR DESCRIPTION
**Pull request by NachoSoto**
### TODO:

- [x] Set environment variable in CircleCI.
~~- [x] [Detect redirects](https://developer.apple.com/documentation/foundation/urlsessiontaskdelegate/1411626-urlsession) and log. Verify this is happening using `TestLogger`.~~

### CODE REVIEW

File: fastlane/Fastfile [hunk 1]
```diff
@@ -246,6 +246,11 @@ platform :ios do
         './Tests/BackendIntegrationTests/Constants.swift',
         './Tests/InstallationTests/CommonFiles/RCInstallationRunner.m'
       ]
+      previous_text: "REVENUECAT_LOAD_SHEDDER_API_KEY",
+      new_text: ENV["REVENUECAT_LOAD_SHEDDER_API_KEY"],
```
Comment (NachoSoto) [inline fastlane/Fastfile:250]: Need to set this in CircleCI once it's available.

File: Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift [hunk 2]
```diff
@@ -44,7 +44,12 @@ class BaseBackendIntegrationTests: XCTestCase {
         // Avoid continuing with potentially bad data after a failed assertion
         self.continueAfterFailure = false
 
-        guard Constants.apiKey != "REVENUECAT_API_KEY", Constants.proxyURL != "REVENUECAT_PROXY_URL" else {
+        let apiKey = self.apiKey
+        let proxyURL = self.proxyURL
+
+        guard apiKey != "REVENUECAT_API_KEY",
+                apiKey != "REVENUECAT_LOAD_SHEDDER_API_KEY",
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:51]: I think this should be `loadShedderApiKey`?

File: fastlane/Fastfile (hunk 1 above)
Comment (tonidero) [reply fastlane/Fastfile:250]: Hmm I think you will need a new call to `replace_text_in_files` here for this replacement?

Comment (NachoSoto) [reply fastlane/Fastfile:250]: What am I doing 🤦🏻

File: Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift (hunk 2 above)
Comment (tonidero) [reply Tests/BackendIntegrationTests/BaseBackendIntegrationTests.swift:51]: Ahh right 👍

File: Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    }
+
+    func testCanGetOfferings() async throws {
+        let receivedOfferings = try await Purchases.shared.offerings()
+        expect(receivedOfferings.all).toNot(beEmpty())
+    }
+
+    func testCanPurchasePackage() async throws {
+        try await self.purchaseMonthlyOffering()
+    }
```
Comment (tonidero) [inline Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift:42]: I guess we will need to add one as well for getting the `ProductEntitlementMapping` as well?

**Review by tonidero — APPROVED** [review]
LGTM!

File: Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift (hunk 3 above)
Comment (NachoSoto) [reply Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift:42]: I did that in #2355 for the normal server. Once that's done I'll add it here too 👍🏻

Comment (NachoSoto) [reply Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift:42]: Done here now.

File: Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testProductEntitlementMapping() async throws {
+        try AvailabilityChecks.iOS15APIAvailableOrSkipTest()
+
+        let result = try await Purchases.shared.productEntitlementMapping()
+        expect(result.entitlementsByProduct).toNot(beEmpty())
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func testProductEntitlementMappingRedirectsToLoadShedder() async throws {
```
Comment (NachoSoto) [inline Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift:67]: @aboedo will probably need to remove this, sounds like the server will actually redirect internally..

File: Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+class LoadShedderStoreKit1IntegrationTests: BaseStoreKitIntegrationTests {
+
+    override var apiKey: String { return Constants.loadShedderApiKey }
+
+    override class var storeKit2Setting: StoreKit2Setting {
+        return .disabled
+    }
+
+    // temporarily disabling signature verification for load shedder until signing is
```
Comment (NachoSoto) [inline Tests/BackendIntegrationTests/LoadShedderIntegrationTests.swift:35]: I like this

File: Tests/TestingApps/PurchaseTester/RevenueCat_IntegrationPurchaseTesterConfiguration.storekit [hunk 6]
```diff
@@ -329,14 +329,48 @@
...
     }
   ],
   "subscriptionOffersKeyPair" : {
     "id" : "C815358F",
     "privateKey" : "MIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgMpw0LScUC+8o+WsEAM20KOMzZOMSWKFjUixtwDidPMugCgYIKoZIzj0DAQehRANCAAQIMw4YEVZrDPrk/1dx5WV3hpkuy8enQs+yDAuJRtsrOnzRH6EFgTyiydprKlgUOHW6xBocrRNTMZXioCBQtchg"
   },
   "version" : {
-    "major" : 1,
-    "minor" : 2
+    "major" : 2,
```
[the remark this answers is not in the document]
Comment (aboedo) [reply Tests/TestingApps/PurchaseTester/RevenueCat_IntegrationPurchaseTesterConfiguration.storekit:373]: I'd missed that, good to know. That being said, it looks like [14.1 will be required starting April 25]([App Store submission requirement starts April 25](https://developer.apple.com/news/?id=jd9wcyov)), so this shouldn't be an issue. Also, this is just for our integration tests

**Comment by aboedo** [comment]
These aren't passing right now because the product was mapped to two entitlements: `pro` and `premium` and tests expect only `premium`. I've updated the mapping, they should start passing correctly when caches update tomorrow.

**Comment by aboedo** [comment]
and finally all tests are passing! 🎉 merging

File: Tests/TestingApps/PurchaseTester/RevenueCat_IntegrationPurchaseTesterConfiguration.storekit (hunk 6 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Tests/TestingApps/PurchaseTester/RevenueCat_IntegrationPurchaseTesterConfiguration.storekit:373]: Yeah, sounds good! I can send a PR to drop support for Xcode 13 this week.

**Pull request merged**